bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/Workbench.java

```diff
@@ -1483,6 +1489,63 @@ IWorkbenchWindow createWorkbenchWindow(IAdaptable input, IPerspectiveDescriptor
 		return result;
 	}
 
+	private void placeNearActiveShell(MWindow window) {
+		if (getDisplay() == null)
+			return;
+
+		Shell activeShell = getDisplay().getActiveShell();
+		if (activeShell == null)
+			return;
+
+		Rectangle newShellBounds = activeShell.getBounds();
+		Monitor currentMonitor = findMonitorThatContains(newShellBounds.x, newShellBounds.y);
+
+		Rectangle currentMonitorBounds = currentMonitor.getBounds();
+
+		final int offsetToExistingShell = 100;
+		newShellBounds.x += offsetToExistingShell;
+		newShellBounds.y += offsetToExistingShell;
+		moveIntoBounds(newShellBounds, currentMonitorBounds);
+		window.setX(newShellBounds.x);
+		window.setY(newShellBounds.y);
+	}
+
+	/**
+	 * Find the monitor that contains a point.
+	 *
+	 * @param x the coordinate "X"
+	 * @param y the coordinate "Y"
+	 * @return the first monitor that contains the point or, if none is found, the
+	 *         primary monitor.
+	 */
+	private Monitor findMonitorThatContains(int x, int y) {
+		for (Monitor m : getDisplay().getMonitors()) {
+			if (m.getBounds().contains(x, y))
+				return m;
+		}
+
+		return getDisplay().getPrimaryMonitor();
+	}
+
+	private static void moveIntoBounds(Rectangle rectangleToMove, Rectangle bounds) {
+		// move into bounds if it's too far to the right
+		if (rectangleToMove.x + rectangleToMove.width > bounds.x + bounds.width)
+			rectangleToMove.x = bounds.x + bounds.width - rectangleToMove.width - 100;
```

Comment:
I find it a little unintuitive that the new window bounces back for some magic padding of the display (100 px), but only if it would leave the display otherwise. In case the initial coordinates of the new window have a padding of only 1 px to the display bounds, it would not bounce back.
In case such a padding is wanted, I would prefer to always ensure it and not only if the window would exceed the display otherwise.
The easiest solution (and most comprehensible anyway) would be to call call this methods with bounds that already reflect the wanted padding rather than having some magic value subtracted within the method.

In terms of code style: please always enclose expression blocks in curly braces, even they only contain one statement. There have already been enough severe bugs by leaving them out 🙂